src/main/kotlin/cacophony/semantic/syntaxtree/SyntaxTree.kt

```diff
@@ -377,150 +380,151 @@ sealed class OperatorBinary(
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is Greater
+        override fun isEquivalent(other: SyntaxTree?): Boolean = super.isEquivalent(other) && other is Greater
     }
 
     class LessEqual(
         range: Pair<Location, Location>,
         lhs: Expression,
         rhs: Expression,
     ) : LogicalOperator(range, lhs, rhs) {
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is LessEqual
+        override fun isEquivalent(other: SyntaxTree?): Boolean = super.isEquivalent(other) && other is LessEqual
     }
 
     class GreaterEqual(
         range: Pair<Location, Location>,
         lhs: Expression,
         rhs: Expression,
     ) : LogicalOperator(range, lhs, rhs) {
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is GreaterEqual
+        override fun isEquivalent(other: SyntaxTree?): Boolean = super.isEquivalent(other) && other is GreaterEqual
     }
 
     class Equals(
         range: Pair<Location, Location>,
         lhs: Expression,
         rhs: Expression,
     ) : LogicalOperator(range, lhs, rhs) {
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is Equals
+        override fun isEquivalent(other: SyntaxTree?): Boolean = super.isEquivalent(other) && other is Equals
     }
 
     class NotEquals(
         range: Pair<Location, Location>,
         lhs: Expression,
         rhs: Expression,
     ) : LogicalOperator(range, lhs, rhs) {
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is NotEquals
+        override fun isEquivalent(other: SyntaxTree?): Boolean = super.isEquivalent(other) && other is NotEquals
     }
 
     class LogicalAnd(
         range: Pair<Location, Location>,
         lhs: Expression,
         rhs: Expression,
     ) : LogicalOperator(range, lhs, rhs) {
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is LogicalAnd
+        override fun isEquivalent(other: SyntaxTree?): Boolean = super.isEquivalent(other) && other is LogicalAnd
     }
 
     class LogicalOr(
         range: Pair<Location, Location>,
         lhs: Expression,
         rhs: Expression,
     ) : LogicalOperator(range, lhs, rhs) {
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is LogicalOr
+        override fun isEquivalent(other: SyntaxTree?): Boolean = super.isEquivalent(other) && other is LogicalOr
     }
 
     class Assignment(
         range: Pair<Location, Location>,
-        lhs: Expression,
+        lhs: Assignable,
         rhs: Expression,
-    ) : OperatorBinary(range, lhs, rhs) {
-        override fun isEquivalent(other: Expression?): Boolean = super.isEquivalent(other) && other is Assignment
+    ) : LValueOperator(range, lhs, rhs), Assignable {
```

Comment:
should this be Assignable?

Reply:
Imo yes, since `a += b` to me returns `a` after having added `b` to it. It makes stuff like `(a += b) *= c` legal (idk if that's good or bad, but I think that's the way it was before my changes, so I decided not to change the behavior

Reply:
ohh I see, it's VariableDeclaration which returns Unit, Assignment returns lhsType. That's right